### PR TITLE
fix:104 unsafe integer conversion in instance_utils.go

### LINKED ISSUE #104: Fix unsafe integer conversion in instance_utils.go
**File**: `pkg/reconciler/instance/utils/instance_utils.go`  
**Line**: 145  
**Rule**: `go/incorrect-integer-conversion`  
**Severity**: High  

### Issue  
The current implementation converts a string to `int` (architecture-dependent size) and then downcasts to `int32` without bounds checking:  
```go
list := strings.Split(pod.Name, "-")
componentId = list[len(list)-1]
id, _ := strconv.Atoi(componentId)  //  ❌ Unsafe conversion
return int32(id)
```  
If the integer parsed from the pod name exceeds `math.MaxInt32` (2147483647) or is below `math.MinInt32` (-2147483648), the conversion to `int32` will silently wrap, causing undefined behavior (e.g., negative IDs, overflow bugs).

### Recommended Fix  
Replace `strconv.Atoi` with **explicit bounds checking** or use `strconv.ParseInt` with a `32`-bit size:  

**Option 1 (ParseInt with bitSize=32):**  
```go
idStr := list[len(list)-1]
id, err := strconv.ParseInt(idStr, 10, 32)  // ✅ Automatically enforces 32-bit bounds
if err != nil {
    return 0 // or handle error appropriately
}
return int32(id)
```  

**Option 2 (Atoi + Manual Bounds Check):**  
```go
idStr := list[len(list)-1]
id, err := strconv.Atoi(idStr)
if err != nil {
    return 0 
}
// Explicit bounds check
if id < math.MinInt32 || id > math.MaxInt32 {
    return 0 // or handle overflow
}
return int32(id)
```  

### Why This Matters  
-   **Security Impact**: Unchecked conversions enable integer overflow exploits (e.g., resource exhaustion, bypassing checks).  
-  📏 **Correctness**: Values outside `int32` range corrupt data silently.  
-  ⚠️ **CodeQL Alert**: This violates [go/incorrect-integer-conversion](https://codeql.github.com/codeql-query-help/go/go-incorrect-integer-conversion/) (reference: `strconv.Atoi` → `int32`).

### PR DESCRIPTION
**Pull request by gzb2025**
### Ⅰ. Motivation

The GetPodComponentID function in instance_utils.go had two main issues:
Unsafe integer conversion: Used strconv.Atoi (which returns int, size-dependent on system) and directly cast to int32, risking overflow on 64-bit systems.
Incorrect negative ID parsing from pod name: When parsing component ID from pod names (e.g., instance-db--987), the original logic failed to recognize negative signs, resulting in incorrect positive values.
These issues could lead to unexpected behavior in component ID resolution, affecting instance management and scaling logic.
### Ⅱ. Modifications

Improved integer parsing for labels:
Replaced strconv.Atoi with strconv.ParseInt(componentId, 10, 32) to explicitly parse values into int32 range, with error handling to return 0 on failure.
Fixed negative ID parsing from pod names:
When labels are missing, instead of directly taking the last segment of strings.Split(pod.Name, "-"), added logic to:
Traverse backward to find the last non-empty segment (handling empty strings from -- splits).
Used strconv.ParseInt with 32 bit size for consistent range checks, returning 0 on parsing errors.
Added comprehensive test cases (in instance_utils_test.go) to cover edge cases.
### Ⅲ. Does this pull request fix one issue?

fixes #104 

### Ⅳ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.
Added TestGetPodComponentID with 12 test scenarios:
Valid positive/negative IDs from labels.
Invalid label values (non-numeric, exceeding int32 max/min).
Invalid pod name values (non-numeric, exceeding int32 max/min).
Boundary values (int32 max: 2147483647, int32 min: -2147483648) in labels and pod names.
These tests verify correct parsing, error handling, and range checks.
### Ⅴ. Describe how to verify it
Unit tests: Run go test ./pkg/reconciler/instance/utils/ -v -run TestGetPodComponentID to ensure all test cases pass.
Manual verification:
Create pods with labels containing valid/invalid component IDs (e.g., component-id: "-123", component-id: "2147483648").
Create pods with names containing negative IDs (e.g., instance-web--456) and non-numeric suffixes (e.g., instance-db-abc).
Call GetPodComponentID on these pods and verify returns match expectations (correct IDs for valid cases, 0 for invalid cases).
## Checklist

- [✅] Format your code `make fmt`.
- [✅] Add unit tests or integration tests.
- [ ] Update the documentation related to the change.

### CODE REVIEW

File: pkg/reconciler/instance/utils/instance_utils_test.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	"math"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/rbgs/api/workloads/v1alpha1"
+)
+
+// TestGetPodComponentID 测试修复后的GetPodComponentID函数
```
Comment (RongGu) [inline pkg/reconciler/instance/utils/instance_utils_test.go]: Please add comments in English.

**Review by RongGu — CHANGES_REQUESTED** [review]

**Review by cheyang — COMMENTED** [review]
## Review Summary

**Status: needs rebase** — this PR currently has merge conflicts against main.

### What works well

The core fix is correct: replacing `strconv.Atoi` followed by an unchecked `int32()` cast with `strconv.ParseInt(componentId, 10, 32)` eliminates the integer overflow vulnerability from issue #104. The error path now returns 0 instead of producing silently-truncated garbage. Test coverage is solid (12 cases covering both paths, boundary values, overflow, and invalid input).

Previous review feedback (English comments, DRY consolidation) has been addressed.

### Concerns

1. **Merge conflict (blocking):** The PR is marked `CONFLICTING`. Please rebase onto the latest `main` and resolve conflicts before this can proceed.

2. **Behavioral change beyond the stated scope:** The negative-ID detection from pod names (treating `--` in a name like `instance-db--987` as encoding `-987`) is new behavior. The old code would have parsed that as `987`. If the project genuinely uses this convention for negative component IDs, this is fine — but it's worth confirming this is intentional rather than speculative. A brief note in the commit message or PR description explaining why negative IDs from pod names are expected would help future readers.

3. **Minor: label value validity.** Kubernetes label values cannot start with `-`, so the negative-label test case (`"-456"`) won't occur in practice. It's harmless to keep for defense-in-depth, just noting it doesn't correspond to a real scenario.

### Next step

Rebase onto `main` to resolve the conflict, then this should be good to go pending maintainer sign-off.

**Comment by cheyang** [comment]
Hey @gzb2025 — this PR has merge conflicts with `main`. Could you rebase onto the latest `main` branch? The fix itself looks correct from the prior review; just the conflict needs resolving. Thanks!

**Comment by cheyang** [comment]
Re-reviewed after rebase on 0fdde15. The merge conflict from round 1 is resolved, and the int32 bounds via strconv.ParseInt(..., 10, 32) remain correct: overflow inputs surface as parse errors and map to -1, matching the documented contract. The DRY consolidation of the two parse paths is a nice cleanup with no behavioral change. Test coverage now includes the int32 max boundary (2147483647), overflow (2147483648), non-numeric values from both the label and the pod-name fallback, and the double-hyphen pod name case. Looks good to merge once CI is green.

**Comment by cheyang** [comment]
@gzb2025 The PR is labeled and looks good code-wise, but the `copyright-check` action is now failing — most likely a license header is missing on one of the modified/added files. Could you check the [job output](https://github.com/sgl-project/rbg/actions/runs/28219768656/job/84262830033) and add the standard copyright header to any flagged file? Once that turns green and the e2e/envtest jobs finish, the auto-merge will pick this up.
